Don't truncate Tuple, Union, and generic types in error messages

This gives the expected error::

    from typing import List, Tuple
    x = [
        ('', [''], 0),
        ]  # type: List[Tuple[str, List[str], str]]

Error::

    z.py, line 2: List item 1 has incompatible type "Tuple[str, List[str], str]"

But this shortens the type to just 'tuple'::

    from typing import List, Tuple
    x = [
        ('', '', [''], 0),
        ]  # type: List[Tuple[str, str, List[str], str]]

Error::

    z.py, line 2: List item 1 has incompatible type tuple

I've been bitten by this as well. In retrospect, truncating long types now seems like a bad idea in general. It might still be occasionally useful in cases like `expected Tuple[...], got int`, where the truncated form would contain enough information about the types to highlight the difference, but the benefits would be marginal and not certain. Maybe just leave out the type truncation -- we can reconsider and add it back if it results in confusing messages.

There are other options such as run length encoding long tuples (for example, `Tuple[str, ..., str] (9 items)` but until we encounter cases where these would be helpful it's uncertain whether these would be worth implementing. 

Another option for cases where we display an expected and an actual type would be to run a more complete type diff algorithm and highlight the differences (e.g., `expected tuple with length 9, got tuple of length 8`). Even then we could display additional lines with the full types. Hypothetical example:

```
x.py:123: error: Argument 1 to "f" has incompatible type tuple (length 9); expected tuple (length 8)
x.py:123: note:     Actual type:   Tuple[int, int, int, int, int, int, int, int, int]
x.py:123: note:     Expected type: Tuple[int, int, int, int, int, int, int, int]
```

By aligning the types it would be easier to see how they are different in case they are similar but not compatible.

--- a/mypy/messages.py
+++ b/mypy/messages.py
@@ -235,7 +235,7 @@ class MessageBuilder:
                 for arg in itype.args:
                     a.append(strip_quotes(self.format(arg)))
                 s = ', '.join(a)
-                if len((base_str + s)) < 25:
+                if len((base_str + s)) < 250:
                     return '{}[{}]'.format(base_str, s)
                 else:
                     return '{}[...]'.format(base_str)

--- a/mypy/messages.py
+++ b/mypy/messages.py
@@ -235,7 +235,7 @@ class MessageBuilder:
                 for arg in itype.args:
                     a.append(strip_quotes(self.format(arg)))
                 s = ', '.join(a)
-                if len((base_str + s)) < 250:
+                if len((base_str + s)) < 150:
                     return '{}[{}]'.format(base_str, s)
                 else:
                     return '{}[...]'.format(base_str)

--- a/mypy/messages.py
+++ b/mypy/messages.py
@@ -250,7 +250,7 @@ class MessageBuilder:
             for t in typ.items:
                 items.append(strip_quotes(self.format(t)))
             s = '"Tuple[{}]"'.format(', '.join(items))
-            if len(s) < 40:
+            if len(s) < 400:
                 return s
             else:
                 return 'tuple(length {})'.format(len(items))

--- a/test-data/unit/check-async-await.test
+++ b/test-data/unit/check-async-await.test
@@ -363,7 +363,7 @@ def plain_host_generator() -> Generator[str, None, None]:
     x = yield from plain_generator()
     x = yield from plain_coroutine()  # E: "yield from" can't be applied to Awaitable[int]
     x = yield from decorated_generator()
-    x = yield from decorated_coroutine()  # E: "yield from" can't be applied to AwaitableGenerator[...]
+    x = yield from decorated_coroutine()  # E: "yield from" can't be applied to AwaitableGenerator[Any, Any, int, Awaitable[int]]
     x = yield from other_iterator()
     x = yield from other_coroutine()  # E: "yield from" can't be applied to "Aw"
 

--- a/test-data/unit/check-generics.test
+++ b/test-data/unit/check-generics.test
@@ -577,7 +577,7 @@ from typing import Callable, TypeVar, Generic
 S = TypeVar('S')
 T = TypeVar('T')
 a = None # type: A[object, Callable[[], None]]
-f(a) # E: Argument 1 to "f" has incompatible type A[...]; expected "B"
+f(a) # E: Argument 1 to "f" has incompatible type A[object, Callable[[], None]]; expected "B"
 
 def f(a: 'B') -> None: pass
 class A(Generic[S, T]): pass

--- a/test-data/unit/check-generics.test
+++ b/test-data/unit/check-generics.test
@@ -533,14 +533,33 @@ D = TypeVar('D')
 E = TypeVar('E')
 F = TypeVar('F')
 G = TypeVar('G')
-class O: pass
-a = None # type: A[object, object, object, object, object, object]
+H = TypeVar('H')
+I = TypeVar('I')
+J = TypeVar('J')
+K = TypeVar('K')
+L = TypeVar('L')
+M = TypeVar('M')
+N = TypeVar('N')
+O = TypeVar('O')
+P = TypeVar('P')
+Q = TypeVar('Q')
+R = TypeVar('R')
+S = TypeVar('S')
+T = TypeVar('T')
+U = TypeVar('U')
+V = TypeVar('V')
+W = TypeVar('W')
+X = TypeVar('X')
+Y = TypeVar('Y')
+Z = TypeVar('Z')
+class OO: pass
+a = None # type: A[object, object, object, object, object, object, object, object, object, object, object, object, object, object, object, object, object, object, object, object, object, object, object, object, object]
 
-f(a) # E: Argument 1 to "f" has incompatible type A[...]; expected "O"
+f(a) # E: Argument 1 to "f" has incompatible type A[...]; expected "OO"
 
-def f(a: O) -> None:
+def f(a: OO) -> None:
     pass
-class A(Generic[B, C, D, E, F, G]): pass
+class A(Generic[B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z]): pass
 
 [case testErrorWithShorterGenericTypeName]
 from typing import TypeVar, Generic

--- a/test-data/unit/check-tuples.test
+++ b/test-data/unit/check-tuples.test
@@ -573,13 +573,13 @@ class A:
 [case testLargeTuplesInErrorMessages]
 
 a = None # type: LongTypeName
-a + (a, a, a, a, a, a, a) # Fail
+a + (a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a) # Fail
 
 class LongTypeName:
     def __add__(self, x: 'LongTypeName') -> 'LongTypeName': pass
 [builtins fixtures/tuple.pyi]
 [out]
-main:3: error: Unsupported operand types for + ("LongTypeName" and tuple(length 7))
+main:3: error: Unsupported operand types for + ("LongTypeName" and tuple(length 50))
 
 
 -- Tuple methods

--- a/test-data/unit/pythoneval-asyncio.test
+++ b/test-data/unit/pythoneval-asyncio.test
@@ -428,7 +428,7 @@ loop.run_until_complete(h())
 loop.close()
 [out]
 _program.py: note: In function "h3":
-_program.py:18: error: Incompatible return value type (got Future[Future[int]], expected Future[...])
+_program.py:18: error: Incompatible return value type (got Future[Future[int]], expected Future[Future[Future[int]]])
 
 [case testErrorOneLessFutureInReturnType]
 import typing

